#' @title Plot Qualimap read alignment 
#' 
#' @description Takes in the list object generated from \code{qualimap()} function and generates 
#' read alignment plots.
#' 
#' @seealso \code{\link{plot_genomic_origin}}, \code{\link{plot_bias_profile}}, \code{\link{plot_coverage_profile}}, \code{\link{plot_junction_analysis}}
#' @param x The list generated by running \code{qualimap()} function that has gathered all tables from all samples together.
#' @param type Package to use to plot the summaries. Only \code{ggplot2} is implemented for now.
#' @param group Default is \code{NULL}. If the plots need to be grouped by samples, then a two 
#' column \code{data.table} has to be provided with names \code{sample_name} and \code{grp}. The 
#' column \code{grp} should have identical values for all samples belonging to the same group.
#' @export
plot_read_alignment <- function(x, type=c("ggplot2", "rbokeh"), group=NULL) {

	dt = data.table:::shallow(x[["Reads alignment"]])[, Type := factor(Type)]
	if (!is.null(group)) dt[group, grp := i.grp, on="sample_name"][, grp := factor(grp)]
	type = match.arg(type)
	switch(type, 
		ggplot2 = {
			if (!require(ggplot2))
				stop("Package 'ggplot2' is not available.")
			pl = ggplot(dt, aes(x=Type, y = Count))
			pl = pl + if (!is.null(group)) geom_boxplot(aes(fill = grp)) else geom_boxplot()
			pl = pl + facet_wrap(~Type, scales="free") + theme_bw()
			pl = pl + scale_x_discrete(breaks=NULL) + scale_fill_brewer(palette="Set1")
			pl = pl + theme(axis.text.x=element_text(angle=45, vjust=1), text=element_text(size=20L))
		}, 
		rboken = {
			stop("Not yet implemented.")
		})
	pl
}

#' @title Plot Qualimap genomic origin
#' 
#' @description Takes in the list object generated from \code{qualimap()} function and generates 
#' genomic origin plots.
#' 
#' @seealso \code{\link{plot_read_alignment}}, \code{\link{plot_bias_profile}}, \code{\link{plot_coverage_profile}}, \code{\link{plot_junction_analysis}}
#' @param x The list generated by running \code{qualimap()} function that has gathered all tables from all samples together.
#' @param type Package to use to plot the summaries. Only \code{ggplot2} is implemented for now.
#' @export
plot_genomic_origin <- function(x, type=c("ggplot2", "rbokeh")) {

	dt = data.table:::shallow(x[["Reads genomic origin"]])
	dt[, Region := factor(Region)][, sample_name := factor(sample_name)]
	type = match.arg(type)
	switch(type, 
		ggplot2 = {
			if (!require(ggplot2))
				stop("Package 'ggplot2' is not available.")
			pl = ggplot(dt, aes(x=sample_name, y=`Read %`, fill=Region)) + 
				    geom_bar(stat="identity", position="stack") + theme_bw() + 
				    theme(axis.text.x=element_text(angle=45, vjust=1), text=element_text(size=20L)) + 
				    scale_fill_brewer(palette="Set1")
		}, 
		rboken = {
			stop("Not yet implemented.")
		})
	pl
}

#' @title Plot Qualimap bias profile
#' 
#' @description Takes in the list object generated from \code{qualimap()} function and generates 
#' bias profile plots.
#' 
#' @seealso \code{\link{plot_read_alignment}}, \code{\link{plot_genomic_origin}}, \code{\link{plot_coverage_profile}}, \code{\link{plot_junction_analysis}}
#' @param x The list generated by running \code{qualimap()} function that has gathered all tables from all samples together.
#' @param type Package to use to plot the summaries. Only \code{ggplot2} is implemented for now.
#' @export
plot_bias_profile <- function(x, type=c("ggplot2", "rbokeh")) {

	dt = data.table:::shallow(x[["Transcript coverage profile"]])
	dt[, Position := factor(Position)][, sample_name := factor(sample_name)]
	type = match.arg(type)
	switch(type, 
		ggplot2 = {
			if (!require(ggplot2))
				stop("Package 'ggplot2' is not available.")
			pl = ggplot(dt, aes(x=sample_name, y=Value)) + 
				    geom_bar(stat="identity") + theme_bw() + 
				    facet_wrap(~ Position, ncol=1L) + 
				    theme(axis.text.x=element_text(angle=45, vjust=1), text=element_text(size=20L))
		}, 
		rboken = {
			stop("Not yet implemented.")
		})
	pl	
}

#' @title Plot Qualimap coverage profile
#' 
#' @description Takes in the list object generated from \code{qualimap()} function and generates 
#' coverage profile plots.
#' 
#' @seealso \code{\link{plot_read_alignment}}, \code{\link{plot_genomic_origin}}, \code{\link{plot_bias_profile}}, \code{\link{plot_junction_analysis}}
#' @param x The list generated by running \code{qualimap()} function that has gathered all tables from all samples together.
#' @param type Package to use to plot the summaries. Only \code{ggplot2} is implemented for now.
#' @export
plot_coverage_profile <- function(x, type=c("ggplot2", "rbokeh")) {

	dt = data.table:::shallow(x[["Total coverage"]])
	dt[, sample_name := factor(sample_name)]
	type = match.arg(type)
	switch(type, 
		ggplot2 = {
			if (!require(ggplot2))
				stop("Package 'ggplot2' is not available.")
			pl = ggplot(dt, aes(x=`#Transcript position`, y=`Transcript coverage profile`)) + 
				    geom_line(aes(colour=sample_name)) + theme_bw() + 
				    theme(axis.text.x=element_text(angle=45, vjust=1), text=element_text(size=20L)) + 
				    scale_colour_brewer(palette="Set1") 
		}, 
		rboken = {
			stop("Not yet implemented.")
		})
	pl
}

#' @title Plot Qualimap junction analysis
#' 
#' @description Takes in the list object generated from \code{qualimap()} function and generates 
#' junction analysis plots.
#' 
#' @seealso \code{\link{plot_read_alignment}}, \code{\link{plot_genomic_origin}}, \code{\link{plot_coverage_profile}}, \code{\link{plot_bias_profile}}
#' @param x The list generated by running \code{qualimap()} function that has gathered all tables from all samples together.
#' @param type Package to use to plot the summaries. Only \code{ggplot2} is implemented for now.
#' @export
plot_junction_analysis <- function(x, type=c("ggplot2", "rbokeh")) {

	dt = data.table:::shallow(x[["Junction analysis"]])
	dt[, sample_name := factor(sample_name)]
	type = match.arg(type)
	switch(type, 
		ggplot2 = {
			if (!require(ggplot2))
				stop("Package 'ggplot2' is not available.")
			pl = ggplot(dt, aes(x=Junction, y=`Read %`)) + 
				    geom_bar(stat="identity") + theme_bw() + 
				    theme(axis.text.x=element_text(angle=45, hjust=1), text=element_text(size=20L)) + 
				    facet_wrap(~ sample_name, ncol=1L)
		}, 
		rboken = {
			stop("Not yet implemented.")
		})
	pl	
}